cpu/stm32/bootloader/reset.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32
+ * @{
+ *
+ * @file
+ * @brief       Trigger reset to the bootloader stored in the internal boot ROM
+ *              memory.
+ *
+ *              This will start the DFU/UART/SPI bootloader.
+ *              See application note AN2606 for which options are available on
+ *              your individual MCU.
+ *
+ * @author      Benjamin Valentin <benpicco@googlemail.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "periph_cpu.h"
+#include "sched.h"
+
+/* repurpose the sched_context_switch_request variable to signal
+ * jump to bootloader after reset */
+#define MAGIC               sched_context_switch_request
+#define BOOTLOADER_MAGIC    0xB007AFFE
+
+/* called by reset_handler_default() before memory is initialized */
+void pre_startup(void)
+{
+    if (MAGIC != BOOTLOADER_MAGIC) {
+        return;
+    }
+
+    /* clear magic */
+    MAGIC = 0;
+
+    /* enable SYSCFG clock */
+#if defined(RCC_APB2ENR_SYSCFGEN)
+    RCC->APB2ENR   = RCC_APB2ENR_SYSCFGEN;
+#elif defined(RCC_APB2ENR_SYSCFGCOMPEN)
+    RCC->APB2ENR   = RCC_APB2ENR_SYSCFGCOMPEN
+#endif
+
+    /* remap ROM at zero */
+#if defined(SYSCFG_MEMRMP_MEM_MODE_0)
+    SYSCFG->MEMRMP = SYSCFG_MEMRMP_MEM_MODE_0;
+#elif defined(SYSCFG_CFGR1_MEM_MODE_0)
+    SYSCFG->CFGR1  = SYSCFG_CFGR1_MEM_MODE_0;
+#endif
+
+    /* load bootloader address into r0 */
+    register uint32_t dst __asm__("r0") = STM32_LOADER_ADDR;
+
+    /* jump to the bootloader */
+    __asm__ volatile("mov sp, %0" :: "r" (dst));
+    __asm__ volatile("mov pc, %0" :: "r" (dst));
```

Comment:
Wouldn't it be easier to read to just have a single two-line ASM block and just specify `STM32_LOADER_ADDR` as input passed via register? Or is it important, that specifically r0 is used?

Also: Please add a `memory` clobber to be safe that the compiler does not move code across the inline assembly. (I think I recently caught GCC doing just that in the inlineable IRQ API of AVR. And if the compiler is sure that the registers used in the inline assembly are not touched by the C code that is moved, I think this is even legal to do.)

Reply:
> Wouldn't it be easier to read to just have a single two-line ASM block and just specify `STM32_LOADER_ADDR` as input passed via register? Or is it important, that specifically r0 is used?

Probably, I'm afraid my inline-ASM foo is lacking. `r0` is not significant.